src/nfa/nfa.cc

```diff
@@ -1354,3 +1562,501 @@ void Mata::Nfa::ShortestWordsMap::update_current_words(LengthWordsPair& act, con
+                   && this_state_transitions_iter->symbol > other_state_transitions_iter->symbol)
+            {
+                ++other_state_transitions_iter;
+            }
+            if (other_state_transitions_iter == other_state_transitions_iter_end) { break; }
+        }
+
+        // check both iterators point to the transitions with same symbol
+        if (this_state_transitions_iter->symbol == other_state_transitions_iter->symbol)
+        {
+            compute_for_same_symbols(*this_state_transitions_iter,
+                                     *other_state_transitions_iter);
+
+            ++this_state_transitions_iter;
+            ++other_state_transitions_iter;
+        }
+    }
+}
+
+void Intersection::compute_preserving_epsilon_transitions()
+{
+    initialize_pairs_to_process();
+
+    if (lhs.trans_empty() || rhs.trans_empty()) { return; }
+
+    while (!pairs_to_process.empty())
+    {
+        pair_to_process = *pairs_to_process.begin();
+        pairs_to_process.erase(pair_to_process);
+
+        // find all transitions that have same symbol for first and the second state in the pair_to_process
+        for (const auto& this_state_transitions: lhs.transitionrelation[pair_to_process.first] )
+        {
+            for (const auto& other_state_transitions: rhs.transitionrelation[pair_to_process.second])
+            {
+                compute_for_state_pair_preserving_eps_trans(this_state_transitions,
+                                                            other_state_transitions);
+            }
+        }
+    }
+}
+
+void Intersection::compute_for_state_pair_preserving_eps_trans(const TransSymbolStates& lhs_state_transitions,
+                                                               const TransSymbolStates& rhs_state_transitions)
+{
+    if (lhs_state_transitions.symbol == epsilon)
+    {
+        compute_for_lhs_state_epsilon_transition(lhs_state_transitions);
+    }
+
+    if (rhs_state_transitions.symbol == epsilon)
+    {
+        compute_for_rhs_epsilon_transition(rhs_state_transitions);
+    }
+
+    if (lhs_state_transitions.symbol == rhs_state_transitions.symbol)
+    {
+        compute_for_same_symbols(lhs_state_transitions, rhs_state_transitions);
+    }
+}
+
+void Intersection::compute_for_rhs_epsilon_transition(const TransSymbolStates& rhs_state_transitions)
+{
+    // create transition from the pair_to_process to all pairs between states to which first transition goes and states to which second one goes
+    TransSymbolStates intersection_transition(rhs_state_transitions.symbol);
+    for (State other_state_to: rhs_state_transitions.states_to)
+    {
+        create_product_state_and_trans(pair_to_process.first, other_state_to, intersection_transition);
+    }
+    add_product_transition(intersection_transition);
+}
+
+void Intersection::compute_for_lhs_state_epsilon_transition(const TransSymbolStates& lhs_state_transitions)
```

Comment:
Functions are quite similar. Could be potentially generalized.

Reply:
They are, and I have thought about this for some time already. Nevertheless, the results of the refactorization were not satisfiable. I will think about it some more, then.

Reply:
It's not high priority, though. My reasoning is, that maybe in future, one would want an interesection that takes `n` automata as parameters, instead of just two. I used similar optimisation in one of our solvers, and it helped quite a lot. The question also is, whether an intersection of arbitrary number of autoata is even feasible for NFA. If not, I would leave it like this.

Reply:
That was my thinking as well. I approach @martinhruska about this. We think that for now, we should only consider two NFAs (as it is currently done) and do not preoccupy ourselves with arbitrary number of automata for these operations.

Still, this is a valid point, and we will come back to this somewhere in the future.

Reply:
I will try to better explain my way of thinking.

The 3 similar functions handle different scenarios, for each we iterate over a different set of states and as a result, compute product state for a different combination of states.

Consider the following:
```cpp
        for (State this_state_to: lhs_state_transitions.states_to)
        {
            for (State other_state_to: rhs_state_transitions.states_to)
            {
                create_product_state_and_trans(this_state_to, other_state_to, intersection_transition);
            }
        }

        for (State this_state_to: lhs_state_transitions.states_to)
        {
            create_product_state_and_trans(this_state_to, pair_to_process.second, intersection_transition);
        }

        for (State other_state_to: rhs_state_transitions.states_to)
        {
            create_product_state_and_trans(pair_to_process.first, other_state_to, intersection_transition);
        }

```

Each time, we use something different inside the for loops, even if the general structure is similar. I cannot very well generalize the functions that simply. So far, I have failed in refactoring this in any way which would allow us to unify these functions into one. Nevertheless, I am open to suggestions.

Reply:
As I mentioned, since we'll limit ourselves to binary intersection, then I think the refactoring is pointless anyway :-).

Reply:
Yeah, I agree at this point. I will keep a note of this locally, but I will keep the current state as it is now.